kv/kvserver: TestLeaseTransferRejectedIfTargetNeedsSnapshot timed out [brittle test]

**Describe the problem**

Seen in a [recent CI run from a PR](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/10819670):

```
=== RUN   TestLeaseTransferRejectedIfTargetNeedsSnapshot/reject-after-revoke=true
panic: test timed out after 14m55s
```

[The log](https://teamcity.cockroachdb.com/repository/download/Cockroach_UnitTests_BazelUnitTests/10819670:id/tmp/_tmp/33f576194975d55328ba8cc9ee712e65/logTestLeaseTransferRejectedIfTargetNeedsSnapshot2754873085/kvservertest.log) is filled with:


```
I230707 11:29:29.093448 21103 kv/kvserver_test/client_raft_helpers_test.go:87 â‹® [T1,n3] 1869  dropping r64 Raft message â€¹1->3 MsgApp Term:6 Log:6/17 Commit:27 Entries:[6/18 EntryNormal <omitted>, 6/19 EntryNormal <omitted>, 6/20 EntryConfChangeV2 v2 v3, 6/21 EntryNormal <omitted>, 6/22 EntryConfChangeV2, 6/23 EntryNormal <omitted>, 6/24 EntryNormal <omitted>, 6/25 EntryNormal <omitted>, 6/26 EntryNormal <omitted>, 6/27 EntryNormal <omitted>]â€º
W230707 11:29:29.911732 17254 server/node.go:1168 â‹® [T1,n1,summaries] 1870  health alerts detected: {Alerts:[{StoreID:1 Category:â€¹METRICSâ€º Description:â€¹ranges.underreplicatedâ€º Value:63}]}
I230707 11:29:30.093870 21103 kv/kvserver_test/client_raft_helpers_test.go:87 â‹® [T1,n3] 1871  dropping r64 Raft message â€¹1->3 MsgApp Term:6 Log:6/17 Commit:27 Entries:[6/18 EntryNormal <omitted>, 6/19 EntryNormal <omitted>, 6/20 EntryConfChangeV2 v2 v3, 6/21 EntryNormal <omitted>, 6/22 EntryConfChangeV2, 6/23 EntryNormal <omitted>, 6/24 EntryNormal <omitted>, 6/25 EntryNormal <omitted>, 6/26 EntryNormal <omitted>, 6/27 EntryNormal <omitted>]â€º
```


The test history doesn't show this particular test being flaky, but it looks unlikely to be related to my changes.

Jira issue: CRDB-29544

The test also flakes immediately under `./dev stress`. I don't think we're stressing these packages very much...

<details>
<summary>Details</summary>

```
--- FAIL: TestLeaseTransferRejectedIfTargetNeedsSnapshot (5.42s)
    test_log_scope.go:167: test logs captured to: /tmp/cockroach/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestLeaseTransferRejectedIfTargetNeedsSnapshot3239800029
    test_log_scope.go:81: use -show-logs to present logs inline
    client_replica_test.go:3073: -- test log scope end --
    stopper.go:92: leaked stopper, created at:
        goroutine 101 [running]:
        runtime/debug.Stack()
                GOROOT/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/util/stop.register(0xc0031945a0)
                github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:48 +0x4a
        github.com/cockroachdb/cockroach/pkg/util/stop.NewStopper({0x0, 0x0, 0x0?})
                github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:219 +0xe7
        github.com/cockroachdb/cockroach/pkg/testutils/testcluster.NewTestCluster({_, _}, _, {{{{0x7d92620, 0xc0030bdc00}, {0x0, 0x0}, {0x0, 0x0}, {0x0, ...}, ...}, ...}, ...})
                github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:261 +0x2dd
        github.com/cockroachdb/cockroach/pkg/testutils/testcluster.StartTestCluster({_, _}, _, {{{{0x7d92620, 0xc0030bdc00}, {0x0, 0x0}, {0x0, 0x0}, {0x0, ...}, ...}, ...}, ...})
                github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:235 +0x65
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestLeaseTransferRejectedIfTargetNeedsSnapshot.func1(0xc00395a000, 0x0)
                github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_replica_test.go:2935 +0x1bc
        github.com/cockroachdb/cockroach/pkg/testutils.RunTrueAndFalse.func1(0x0?)
                github.com/cockroachdb/cockroach/pkg/testutils/subtest.go:23 +0x24
        testing.tRunner(0xc00395a000, 0xc0004fe720)
                GOROOT/src/testing/testing.go:1446 +0x10b
        created by testing.(*T).Run
                GOROOT/src/testing/testing.go:1493 +0x35f

```

</details>

--- a/pkg/kv/kvserver/client_replica_test.go
+++ b/pkg/kv/kvserver/client_replica_test.go
@@ -3003,10 +3003,17 @@ func TestLeaseTransferRejectedIfTargetNeedsSnapshot(t *testing.T) {
 		// Replica.AdminTransferLease.
 		transferErrC := make(chan error, 1)
 		if rejectAfterRevoke {
-			_ = tc.Stopper().RunAsyncTask(ctx, "transfer lease", func(ctx context.Context) {
+			require.NoError(t, tc.Stopper().RunAsyncTask(ctx, "transfer lease", func(ctx context.Context) {
 				transferErrC <- tc.TransferRangeLease(*repl0.Desc(), tc.Target(2))
-			})
-			<-transferLeaseReqBlockedC
+			}))
+			select {
+			case <-transferLeaseReqBlockedC:
+			// Expected case: lease transfer triggered our interceptor and is now
+			// waiting there for transferLeaseReqUnblockedCh.
+			case err := <-transferErrC:
+				// Unexpected case: lease transfer errored out before making it into the filter.
+				t.Fatalf("transferErrC unexpectedly signaled: %v", err)
+			}
 		}
 
 		// Truncate the log at index+1 (log entries < N are removed, so this

--- a/pkg/kv/kvserver/client_replica_test.go
+++ b/pkg/kv/kvserver/client_replica_test.go
@@ -2933,6 +2933,9 @@ func TestLeaseTransferRejectedIfTargetNeedsSnapshot(t *testing.T) {
 			ServerArgs: base.TestServerArgs{
 				Knobs: base.TestingKnobs{
 					Store: &kvserver.StoreTestingKnobs{
+						// If we're testing the below-raft check, disable the above-raft check.
+						// See: https://github.com/cockroachdb/cockroach/pull/107526
+						DisableAboveRaftLeaseTransferSafetyChecks: rejectAfterRevoke,
 						TestingRequestFilter: func(ctx context.Context, ba *kvpb.BatchRequest) *kvpb.Error {
 							if rejectAfterRevoke && ba.IsSingleTransferLeaseRequest() {
 								transferLeaseReqBlockOnce.Do(func() {

--- a/pkg/kv/kvserver/replica_range_lease.go
+++ b/pkg/kv/kvserver/replica_range_lease.go
@@ -949,7 +949,7 @@ func (r *Replica) AdminTransferLease(
 		raftStatus := r.raftStatusRLocked()
 		raftFirstIndex := r.raftFirstIndexRLocked()
 		snapStatus := raftutil.ReplicaMayNeedSnapshot(raftStatus, raftFirstIndex, nextLeaseHolder.ReplicaID)
-		if snapStatus != raftutil.NoSnapshotNeeded && !bypassSafetyChecks {
+		if snapStatus != raftutil.NoSnapshotNeeded && !bypassSafetyChecks && !r.store.cfg.TestingKnobs.DisableAboveRaftLeaseTransferSafetyChecks {
 			r.store.metrics.LeaseTransferErrorCount.Inc(1)
 			log.VEventf(ctx, 2, "not initiating lease transfer because the target %s may "+
 				"need a snapshot: %s", nextLeaseHolder, snapStatus)

--- a/pkg/kv/kvserver/testing_knobs.go
+++ b/pkg/kv/kvserver/testing_knobs.go
@@ -225,6 +225,10 @@ type StoreTestingKnobs struct {
 	// leadership when it diverges from the range's leaseholder. This can
 	// also be set via COCKROACH_DISABLE_LEADER_FOLLOWS_LEASEHOLDER.
 	DisableLeaderFollowsLeaseholder bool
+	// If set, the above-raft lease transfer safety checks (that verify that
+	// we don't transfer leases to followers that need a snapshot, etc) are
+	// disabled. The proposal-time checks are not affected by this knob.
+	DisableAboveRaftLeaseTransferSafetyChecks bool
 	// DisableRefreshReasonNewLeader disables refreshing pending commands when a new
 	// leader is discovered.
 	DisableRefreshReasonNewLeader bool

